src/tree_cmd.cpp

```diff
@@ -726,86 +782,126 @@ static void TileLoop_Trees(TileIndex tile)
+						GetTreeType(tile) != TREE_CACTUS &&
+						GetTropicZone(tile) == TROPICZONE_DESERT) {
+					AddTreeGrowth(tile, 1);
+				} else {
+					switch (GB(Random(), 0, 3)) {
+						case 0: // start destructing
+							AddTreeGrowth(tile, 1);
+							break;
 
-						/* Don't plant trees, if ground was freshly cleared */
-						if (IsTileType(tile, MP_CLEAR) && GetClearGround(tile) == CLEAR_GRASS && GetClearDensity(tile) != 3) return;
+						case 1: // add a tree
+							spread = can_plant_extra;
+							if (count < 4 && spread) {
+								AddTreeCount(tile, 1);
+								SetTreeGrowth(tile, 0);
+								spread = false;
+							}
+							break;
 
-						PlantTreesOnTile(tile, treetype, 0, 0);
+						case 2: { // add a neighbouring tree
+							spread = can_plant_extra;
+							break;
+						}
 
-						break;
+						default:
+							return;
 					}
-
-					default:
-						return;
 				}
-			}
-			break;
+				break;
 
-		case 6: // final stage of tree destruction
-			if (!CanPlantExtraTrees(tile)) {
-				/* if trees can't spread just plant a new one to prevent deforestation */
-				SetTreeGrowth(tile, 0);
-			} else if (GetTreeCount(tile) > 1) {
-				/* more than one tree, delete it */
-				AddTreeCount(tile, -1);
-				SetTreeGrowth(tile, 3);
-			} else {
-				/* just one tree, change type into MP_CLEAR */
-				switch (GetTreeGround(tile)) {
-					case TREE_GROUND_SHORE: MakeShore(tile); break;
-					case TREE_GROUND_GRASS: MakeClear(tile, CLEAR_GRASS, GetTreeDensity(tile)); break;
-					case TREE_GROUND_ROUGH: MakeClear(tile, CLEAR_ROUGH, 3); break;
-					case TREE_GROUND_ROUGH_SNOW: {
-						uint density = GetTreeDensity(tile);
-						MakeClear(tile, CLEAR_ROUGH, 3);
-						MakeSnow(tile, density);
-						break;
-					}
-					default: // snow or desert
-						if (_settings_game.game_creation.landscape == LT_TROPIC) {
-							MakeClear(tile, CLEAR_DESERT, GetTreeDensity(tile));
-						} else {
-							uint density = GetTreeDensity(tile);
-							MakeClear(tile, CLEAR_GRASS, 3);
-							MakeSnow(tile, density);
-						}
-						break;
+			case 6: // fina)l stage of tree destruction
+				if (!can_plant_extra) {
+					/* if trees can't spread just plant a new one to prevent deforestation */
+					SetTreeGrowth(tile, 0);
+				} else if (count > 1) {
+					/* more than one tree, delete it */
+					AddTreeCount(tile, -1);
+					SetTreeGrowth(tile, 3);
+				} else {
+					/* no more trees, mark the tile for clearing */
+					clear = true;
 				}
+				break;
+
+			default:
+				AddTreeGrowth(tile, 1);
+				break;
+		}
+	}
+
+	if (clear) {
+		/* All trees died, change type into MP_CLEAR */
+		switch (GetTreeGround(tile)) {
+			case TREE_GROUND_SHORE: MakeShore(tile); break;
+			case TREE_GROUND_GRASS: MakeClear(tile, CLEAR_GRASS, GetTreeDensity(tile)); break;
+			case TREE_GROUND_ROUGH: MakeClear(tile, CLEAR_ROUGH, 3); break;
+			case TREE_GROUND_ROUGH_SNOW: {
+				uint density = GetTreeDensity(tile);
```

Comment:
This could be moved up and used for `case TREE_GROUND_GRASS`

Reply:
Moving it up would mean an extra map array access even in cases when density is not used. Unless you believe compiler to be smart enough to optimize it back, I sure don't :p.
Also, this is not a new code, I just moved it around.